hy/compiler.py

```diff
@@ -776,14 +714,42 @@ def hy_eval(
         import_stdlib=import_stdlib,
     )
 
-    if ast_callback:
-        ast_callback(_ast, expr)
+    if globals is None:
+        globals = module.__dict__
 
     # Two-step eval: eval() the body of the exec call
-    eval(ast_compile(_ast, filename, "exec"), module.__dict__, locals)
+    eval(ast_compile(_ast, filename, "exec"), globals, locals)
 
     # Then eval the expression context and return that
-    return eval(ast_compile(expr, filename, "eval"), module.__dict__, locals)
+    return eval(ast_compile(expr, filename, "eval"), globals, locals)
+
+
+def hy_eval_user(model, globals = None, locals = None, module = None):
+    # This function is advertised as `hy.eval`.
+    """An equivalent of Python's :func:`eval` for evaluating Hy code. The chief difference is that the first argument should be a :ref:`model <models>` rather than source text. If you have a string of source text you want to evaluate, convert it to a model first with :hy:func:`hy.read` or :hy:func:`hy.read-many`::
+
+        (hy.eval '(+ 1 1))             ; => 2
+        (hy.eval (hy.read "(+ 1 1)"))  ; => 2
+
+    The optional arguments ``globals`` and ``locals`` work as in the case of ``eval``. There's one more optional argument, ``module``, which can be a module object or a string naming a module. The module's ``__dict__`` attribute can fill in for ``globals``, if ``module`` is provided but ``globals`` isn't, but the primary purpose of ``module`` is to control where macro calls are looked up. Without this argument, the calling module of ``hy.eval`` is used instead. ::
+
+        (defmacro my-test-mac [] 3)
+        (hy.eval '(my-test-mac))                 ; => 3
+        (import hyrule)
+        (hy.eval '(my-test-mac) :module hyrule)  ; NameError
+        (hy.eval '(list-n 3 1) :module hyrule)   ; => [1 1 1]"""
+
+    value = hy_eval(
+        hytree = model,
+        locals = (inspect.getargvalues(inspect.stack()[1][0]).locals
+            if locals is None
+            else locals),
+        globals = globals,
+        module = get_compiler_module(module, None, True))
+    if locals:
+        # Remove the implicitly added `hy`.
+        del locals['hy']
```

Comment:
So something like `(hy.eval '(hy.repr "foo") …)` works. I already had a test of this, but it was insufficiently stringent (it passed even with `import_stdlib = False`), so I've tightened it. You should now be able to see it fail with `import_stdlib = False` added and this `del` statement commented out.

Reply:
I think `hy` being in the namespace should be handled by inserting it into the `__builtins__` key rather than adding/deleting from `locals`.

From Python `eval`/`exec`:
> If the globals dictionary is present and does not contain a value for the key __builtins__, a reference to the dictionary of the built-in module [builtins](https://docs.python.org/3/library/builtins.html#module-builtins) is inserted under that key before expression is parsed.

Since we're already using the same spec, we can piggyback off of that:
```python
    if globals is not None and '__builtins__' not in globals:
        globals['__builtins__'] = dict(vars(builtins), hy=hy)
```

Reply:
But that would be a needless inconsistency with the way `hy` is usually brought into scope. A normal Hy program starts with an implicit `import hy`; why should a form evaluated with `eval` be any different?

Reply:
In my mind the implicit `import hy` is an implementation detail because there's no other way for us to get `hy` into a bytecode compiled file, and going via `__builtins__` is more consistent with the intention of `hy` existing as an always-accessible name.

Reply:
In my mind it's on purpose, which is why I documented it in `semantics.rst`. I don't know whether futzing with `__builtins__` is in general better or worse than just saying `(import hy)`, but it seems we ought to do the same thing (viz., making `hy` available) in the same way in both cases (viz., running an ordinary program and calling `hy.eval`) so long as we can.

Reply:
Hm. Deleting entries off of `locals` somehow just feels more fraught to me than futzing with `__builtins__`, but as long as it works and isn't causing problems I don't have any reason to insist we use builtins over local/import.

Reply:
Okay. Maybe we'll find a bug eventually. If we then need to switch `hy.eval` to editing `__builtins__`, we can hopefully change the usual case of compiling Hy code to edit `__builtins__` too instead of doing a regular import.

Reply:
Ok bit of a contrived example, but it might not be that much of a stretch if instead of string literals, you were reading and executing snippets of intermixed hy and python out of a file or something:

```python
my_globals = {}
exec("import hy", my_globals)

hy.eval(hy.read("""(do (import types [SimpleNamespace])
                       (setv obj (SimpleNamespace)
                             obj.some-name! 5))"""), my_globals)

print(eval("getattr(obj, hy.mangle('some-name!'))", my_globals))
```

Reply:
I see, so this suggests to me that if `hy` already exists in the `locals` object, I should save it and then assign it back after `hy_eval` returns instead of deleting it. Do you think that suffices?

Reply:
Yeah that should suffice. Technically we'd still have a problem with something like
```python
my_globals = {}
hy.eval(hy.read("(import hy)"), my_globals)
assert "hy" in my_globals
```
but... eh :shipit: